### PR TITLE
Updating maintenance metrics workflow to support overwrite

### PR DESCRIPTION
**Pull request by manasaV3**
## Description
Addresses https://github.com/chanzuckerberg/napari-hub/issues/1199

### Changes introduced
- Adds expiry to the `github-activity` dynamo table infrastructure
- Updates GitHubActivityType to include expiry formatter by type
- Updates query to fetch data for the last 14 months if `type=MONTH` for commit activity
- Updates data accumulator for `type=MONTH` to fill months that don't have an entry with count=0
- Cleans up tests

### CODE REVIEW

File: data-workflows/activity/snowflake_adapter.py [hunk 1]
```diff
@@ -137,66 +173,85 @@ def _cursor_to_timestamp_by_name_mapper(accumulator: dict[str, datetime], cursor
...
-    cursor record and added to the accumulator keyed on name.
-    The cursor contains the fields name, timestamp, and count
-    :param dict[str, List] accumulator: Accumulator that will be updated with new data
-    :param SnowflakeCursor cursor:
-    :returns: Accumulator after data from cursor has been added
-   """
-    for name, timestamp, count in cursor:
-        accumulator.setdefault(name, []).append({'timestamp': timestamp, 'count': count})
-    return accumulator
+def _cursor_to_plugin_activity_mapper(
```
Comment (aganders3) [inline data-workflows/activity/snowflake_adapter.py:176]: The name here was a little confusing (where used above) as this now returns a callable. Maybe `_get_cursor_to_plugin_activity_mapper` or `_make_cursor_to_plugin_activity_mapper`?

File: data-workflows/activity/tests/test_github_activity_model.py [hunk 2]
```diff
@@ -1,30 +1,26 @@
...
-    REPO1: datetime.strptime("05/16/2023 10:24:20", "%m/%d/%Y %H:%M:%S"),
-    REPO2: datetime.strptime("06/26/2023 20:30:00", "%m/%d/%Y %H:%M:%S"),
-}
-FORMATTED_PLUGIN_BY_TS = {
-    repo: ts.replace(day=1) for repo, ts in PLUGIN_BY_EARLIEST_TS.items()
-}
 
 
-def generate_expected(data, granularity, type_id, ts_formatter):
+def generate_expected(data, granularity, type_id, ts_formatter, include_expiry=False):
```
Comment (aganders3) [inline data-workflows/activity/tests/test_github_activity_model.py:18]:
```suggestion
def generate_expected(data, granularity, type_id, ts_formatter, *, include_expiry=False):
```
nitpick (super-nit as this is in test code): I like boolean args to be keyword-only for clarity at the call site

**Review by aganders3 — COMMENTED** [review]
I have a couple quick questions/comments on this but mostly looks okay and tests all pass.

As a small request - it looks like you're running a code-formatter in some of these PRs. I totally support this, but can you possibly try to split those changes into a separate commit? It makes it harder to read for actual changes especially as I am not so intimately familiar with this codebase.

File: data-workflows/activity/github_activity_model.py [hunk 3]
```diff
@@ -1,72 +1,85 @@
...
         lambda timestamp: None,
         "TOTAL:{repo}",
         "COUNT(*) AS commit_count",
-        "name"
+        "name",
+        lambda timestamp: None,
     )
 
-    def format_to_timestamp(self, timestamp: datetime) -> Union[int, None]:
+    def format_to_timestamp(self, timestamp: Optional[date]) -> Optional[int]:
```
[the remark this answers is not in the document]
Comment (manasaV3) [reply data-workflows/activity/github_activity_model.py:57]: > It's a little hard to trace through here but I don't think this arg can be None. For example it gets passed directly to to_datetime in at least some of the formatters, which will fail. I think this is only an annotation issue though.

In the cases such as when we are fetching `TOTAL`, we end up having `None` passed in for the `timestamp` value.

**Comment by manasaV3** [comment]
> As a small request - it looks like you're running a code-formatter in some of these PRs. I totally support this, but can you possibly try to split those changes into a separate commit (doesn't need to be a separate PR)? It makes it harder to read for actual changes especially as I am not so intimately familiar with this codebase.

Sorry, that it has made it hard to review. I had set the IDE to reformat the file on saving, making it harder to move all formatting changes to a separate commit. I will disable that going forward.

**Comment by aganders3** [comment]
> Sorry, that it has made it hard to review. I had set the IDE to reformat the file on saving, making it harder to move all formatting changes to a separate commit. I will disable that going forward.

No worries! Maybe we can just do one bulk PR first that applies formatting only to all files. Then you can leave on this setting in your IDE to maintain the format for new changes. If this gets in the way for history (blame) you can add that one commit to a `.git-blame-ignore-revs` file in the root dir.

**Pull request merged**